router/dataplane.go

```diff
@@ -688,13 +699,17 @@ func (d *DataPlane) AddNextHopBFD(ifID uint16, src, dst *net.UDPAddr, cfg contro
 // @ requires  d.mtx.LockInv() == MutexInvariant!<d!>;
 // @ requires  acc(ctx.Mem(), _)
 // @ requires  acc(&d.mtx, _)
-func (d *DataPlane) Run(ctx context.Context) error {
+// contracts for IO-spec
+// @ requires dp.Valid()
```

Comment:
> this spec does not relate at all the values of `d` with the value of `dp`. Won't we need this later? Whether we can specify this or not is a different question. In particular, because we only have `MutexInvariant!<d!>` after acquiring the lock, we may not be able to specify this in the pre, and we may have to add this as an assumption, in the same way I did later with `d.WellConfigured()` (you can find an explanation in the body of `Run`)

Yes at the moment `d` and `dp` are not related. I left that for now because I don't know how to do that yet. However a connection has to be established at some point because it is needed in `process()`. 


> Furthermore. I think that `dp` should not be passed as a parameter. Instead, it can be computed from d directly. For this, I will need to change the spec of `Run` slightly, to allow to specify these things. Do you think that is reasonable?

I thought the whole reason why we need `dp` is because we require knowledge that is not present in `d`. If this is the case how can be compute `dp` from `d`?

Reply:
Good point. I guess that what we want here is not to compute `dp` from `d`, but that they are compatible, i.e., for the fields that are present in `d`, they should match the values of `dp`.

Reply:
I think that this is what we have to require/assume in `Run`, and we will need to propagate this information to the rest of the data plane

Reply:
Yes but we need to make sure that as long as "d.running == false", `dp` and `d` maintain their equality. For me it is not obvious how to achieve that.

Reply:
Why? To me, this is essentially a precondition of the `Run` method.

Reply:
(we may not be able to write this postcondition due to permission shenanigans at the moment, but that is a different matter)

Reply:
Let's say we have a precondition `d == dp`. As long as `d.running == false` new interfaces could be added to `d` which invalides `d == dp`?

Reply:
I don't think so: I believe that all setters that could invalidate this condition require permission to `acc(&d.running, 1/2)`, which the `Run` method also requires. Thus, we could not call any setter in parallel with `Run` to inviolate this. There is however a Catcha: I think there is a bug in `AddLinkType` that does not check whether the router is running, and thus, does not require this permission (even though we could still require it). I even reported a bug: https://github.com/scionproto/scion/issues/4282

I think that the assumption should be safe to make, as long as we confirm that really all public setters that can invalidate this condition also require `acc(&d.running, 1/2)`

Reply:
If two functions only need half the permission why can't they run in parallel? 
Regardless this bug is a problem because I need `Linktype`.